va/validation-authority.go

```diff
@@ -417,67 +496,138 @@ func (va *ValidationAuthorityImpl) validateDvsni(identifier core.AcmeIdentifier,
+			Type:   core.MalformedProblem,
+			Detail: "Identifier type for HTTP validation was not DNS",
+		}
+
+		va.log.Debug(fmt.Sprintf("%s [%s] Identifier failure", challenge.Type, identifier))
 		return challenge, challenge.Error
 	}
 
-	// Make a connection with SNI = nonceName
-	portString := fmt.Sprintf("%d", va.dvsniPort)
-	hostPort := net.JoinHostPort(addr.String(), portString)
-	challenge.ValidationRecord[0].Port = portString
-	va.log.Notice(fmt.Sprintf("DVSNI [%s] Attempting to validate DVSNI for %s %s",
-		identifier, hostPort, ZName))
-	conn, err := tls.DialWithDialer(&net.Dialer{Timeout: validationTimeout}, "tcp", hostPort, &tls.Config{
-		ServerName:         ZName,
-		InsecureSkipVerify: true,
-	})
+	// Validate that the challenge/response pair is well-formed
+	// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
+	var authorizedKey core.AuthorizedKey
+	err := json.Unmarshal(challenge.AuthorizedKey, &authorizedKey)
+	if err != nil {
+		challenge.Error = &core.ProblemDetails{
+			Type:   core.MalformedProblem,
+			Detail: "AuthorizedKey object on challenge failed to parse as JSON",
+		}
+		va.log.Debug(fmt.Sprintf("%s [%s] HTTP failure: %s", challenge.Type, identifier, err))
+		challenge.Status = core.StatusInvalid
+		return challenge, err
+	}
+	if !authorizedKey.Match(challenge.Token, challenge.AccountKey) {
+		err = fmt.Errorf("Improper token / authorized key pair")
+		challenge.Error = &core.ProblemDetails{
+			Type:   core.MalformedProblem,
+			Detail: err.Error(),
+		}
+		va.log.Debug(fmt.Sprintf("%s [%s] HTTP failure: %s", challenge.Type, identifier, err))
+		challenge.Status = core.StatusInvalid
+		return challenge, err
+	}
+
+	// Perform the fetch
+	path := fmt.Sprintf(".well-known/acme-challenge/%s", challenge.Token)
+	body, challenge, err := va.fetchHTTP(identifier, path, false, challenge)
+	if err != nil {
+		return challenge, err
+	}
 
+	// Parse body as an authorized key object
+	var serverAuthorizedKey core.AuthorizedKey
+	err = json.Unmarshal(body, &serverAuthorizedKey)
 	if err != nil {
+		err = fmt.Errorf("Error parsing authorized key file: %s", err.Error())
+		va.log.Debug(err.Error())
 		challenge.Status = core.StatusInvalid
 		challenge.Error = &core.ProblemDetails{
-			Type:   parseHTTPConnError(err),
-			Detail: "Failed to connect to host for DVSNI challenge",
+			Type:   core.UnauthorizedProblem,
+			Detail: err.Error(),
 		}
-		va.log.Debug(fmt.Sprintf("DVSNI [%s] TLS Connection failure: %s", identifier, err))
 		return challenge, err
 	}
-	defer conn.Close()
 
-	// Check that ZName is a dNSName SAN in the server's certificate
-	certs := conn.ConnectionState().PeerCertificates
-	if len(certs) == 0 {
+	// Check that the account key for this challenge is authorized by this object
+	if !serverAuthorizedKey.MatchAuthorizedKey(authorizedKey) {
+		err = fmt.Errorf("The authorizated keys file from the server did not match this challenge")
+		va.log.Debug(err.Error())
```

Comment:
I think these lines were a testing artefact.  I've changed the content of `err` to have the JSON form of the two authorized key objects.